SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift

```diff
@@ -0,0 +1,160 @@
+//===--- ReleaseDevirtualizer.swift - Devirtualizes release-instructions --===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Devirtualizes release instructions which are known to destruct the object.
+///
+/// This means, it replaces a sequence of
+///    %x = alloc_ref [stack] $X
+///      ...
+///    strong_release %x
+///    dealloc_stack_ref %x
+/// with
+///    %x = alloc_ref [stack] $X
+///      ...
+///    set_deallocating %x
+///    %d = function_ref @dealloc_of_X
+///    %a = apply %d(%x)
+///    dealloc_stack_ref %x
+///
+/// The optimization is only done for stack promoted objects because they are
+/// known to have no associated objects (which are not explicitly released
+/// in the deinit method).
+let releaseDevirtualizerPass = FunctionPass(
+  name: "release-devirtualizer", { function, context in
+    for block in function.blocks {
+      // The last `release_value`` or `strong_release`` instruction before the
+      // deallocation.
+      var lastRelease: RefCountingInst?
+
+      for instruction in block.instructions {
+        if let release = lastRelease {
+          // We only do the optimization for stack promoted object, because for
+          // these we know that they don't have associated objects, which are
+          // _not_ released by the deinit method.
+          if let deallocStackRef = instruction as? DeallocStackRefInst {
+            devirtualizeReleaseOfObject(context, release, deallocStackRef)
+            lastRelease = nil
+            continue
+          }
+        }
+
+        if instruction is ReleaseValueInst || instruction is StrongReleaseInst {
+          lastRelease = instruction as? RefCountingInst
+        } else if instruction.mayReleaseOrReadRefCount {
+          lastRelease = nil
+        }
+      }
+    }
+  }
+)
+
+/// Devirtualize releases of array buffers.
+private func devirtualizeReleaseOfObject(
+  _ context: PassContext,
+  _ release: RefCountingInst,
+  _ deallocStackRef: DeallocStackRefInst
+) {
+  let allocRefInstruction = deallocStackRef.allocRef
+  var root = release.operands[0].value
+  while let newRoot = stripRCIdentityPreservingInsts(root) {
+    root = newRoot
+  }
+
+  guard root === allocRefInstruction else {
+    return
+  }
+
+  createDeallocCall(context, allocRefInstruction.type, release, allocRefInstruction)
+}
+
+/// Replaces the release-instruction `release` with an explicit call to
+/// the deallocating destructor of `type` for `object`.
+private func createDeallocCall(
```

Comment:
you can merge this function with `tryDevirtualizeReleaseOfObject`. IMO, `createDeallocCall` is confusing, because it can fail (if the destructor of the type is not found).

Or: move the `guard let dealloc...` into `tryDevirtualizeReleaseOfObject`